# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/fs/api/FileSystemSyncAccessHandle.cpp
# Commit: b68e19021833
# Full Hash: b68e1902183357e129a1a53490447fc3cc219f4f
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2023-02-07 16:42:31
# Regressor Bug: 1809064
# File Overlap Count: 1
# Description:
#   Bug 1809064 - Add PFileSystemAccessHandleControl top level protocol; r=dom-storage-reviewers,jari
#   
#   Sync access handle methods need to run sync loops to block the thread until a
#   result is available. The main protocol PFileSystemManager and its subprotocol
#   PFileSystemAccessHandle is bound to the main worker event queue, so incoming
# ==============================================================================

diff -r fc3ab5ee8ef3 -r b68e19021833 dom/fs/api/FileSystemSyncAccessHandle.cpp
--- a/dom/fs/api/FileSystemSyncAccessHandle.cpp	Tue Feb 07 08:21:15 2023 +0000
+++ b/dom/fs/api/FileSystemSyncAccessHandle.cpp	Tue Feb 07 08:27:06 2023 +0000
@@ -13,6 +13,7 @@
 #include "mozilla/MozPromise.h"
 #include "mozilla/TaskQueue.h"
 #include "mozilla/dom/FileSystemAccessHandleChild.h"
+#include "mozilla/dom/FileSystemAccessHandleControlChild.h"
 #include "mozilla/dom/FileSystemHandleBinding.h"
 #include "mozilla/dom/FileSystemLog.h"
 #include "mozilla/dom/FileSystemManager.h"
@@ -94,11 +95,14 @@
 FileSystemSyncAccessHandle::FileSystemSyncAccessHandle(
     nsIGlobalObject* aGlobal, RefPtr<FileSystemManager>& aManager,
     mozilla::ipc::RandomAccessStreamParams&& aStreamParams,
-    RefPtr<FileSystemAccessHandleChild> aActor, RefPtr<TaskQueue> aIOTaskQueue,
+    RefPtr<FileSystemAccessHandleChild> aActor,
+    RefPtr<FileSystemAccessHandleControlChild> aControlActor,
+    RefPtr<TaskQueue> aIOTaskQueue,
     const fs::FileSystemEntryMetadata& aMetadata)
     : mGlobal(aGlobal),
       mManager(aManager),
       mActor(std::move(aActor)),
+      mControlActor(std::move(aControlActor)),
       mIOTaskQueue(std::move(aIOTaskQueue)),
       mStreamParams(std::move(aStreamParams)),
       mMetadata(aMetadata),
@@ -111,6 +115,8 @@
   // FileSystemSyncAccessHandle::Create fails, in which case the not yet
   // fully constructed FileSystemSyncAccessHandle is being destroyed.
   mActor->SetAccessHandle(this);
+
+  mControlActor->SetAccessHandle(this);
 }
 
 FileSystemSyncAccessHandle::~FileSystemSyncAccessHandle() {
@@ -125,13 +131,24 @@
     mozilla::ipc::RandomAccessStreamParams&& aStreamParams,
     mozilla::ipc::ManagedEndpoint<PFileSystemAccessHandleChild>&&
         aAccessHandleChildEndpoint,
+    mozilla::ipc::Endpoint<PFileSystemAccessHandleControlChild>&&
+        aAccessHandleControlChildEndpoint,
     const fs::FileSystemEntryMetadata& aMetadata) {
+  WorkerPrivate* const workerPrivate = GetCurrentThreadWorkerPrivate();
+  MOZ_ASSERT(workerPrivate);
+
   auto accessHandleChild = MakeRefPtr<FileSystemAccessHandleChild>();
 
   QM_TRY(MOZ_TO_RESULT(
       aManager->ActorStrongRef()->BindPFileSystemAccessHandleEndpoint(
           std::move(aAccessHandleChildEndpoint), accessHandleChild)));
 
+  auto accessHandleControlChild =
+      MakeRefPtr<FileSystemAccessHandleControlChild>();
+
+  aAccessHandleControlChildEndpoint.Bind(accessHandleControlChild,
+                                         workerPrivate->ControlEventTarget());
+
   QM_TRY_UNWRAP(auto streamTransportService,
                 MOZ_TO_RESULT_GET_TYPED(nsCOMPtr<nsIEventTarget>,
                                         MOZ_SELECT_OVERLOAD(do_GetService),
@@ -143,7 +160,7 @@
 
   RefPtr<FileSystemSyncAccessHandle> result = new FileSystemSyncAccessHandle(
       aGlobal, aManager, std::move(aStreamParams), std::move(accessHandleChild),
-      std::move(ioTaskQueue), aMetadata);
+      std::move(accessHandleControlChild), std::move(ioTaskQueue), aMetadata);
 
   auto autoClose = MakeScopeExit([result] {
     MOZ_ASSERT(result->mState == State::Initial);
@@ -151,9 +168,6 @@
     result->mActor->SendClose();
   });
 
-  WorkerPrivate* const workerPrivate = GetCurrentThreadWorkerPrivate();
-  MOZ_ASSERT(workerPrivate);
-
   workerPrivate->AssertIsOnWorkerThread();
 
   RefPtr<StrongWorkerRef> workerRef = StrongWorkerRef::Create(
@@ -210,8 +224,23 @@
 
   if (mActor) {
     PFileSystemAccessHandleChild::Send__delete__(mActor);
+
+    // `PFileSystemAccessHandleChild::Send__delete__` is supposed to call
+    // `FileSystemAccessHandleChild::ActorDestroy` which in turn calls
+    // `FileSystemSyncAccessHandle::ClearActor`, so `mActor` should be be null
+    // at this point.
     MOZ_ASSERT(!mActor);
   }
+
+  if (mControlActor) {
+    mControlActor->Close();
+
+    // `FileSystemAccessHandleControlChild::Close` is supposed to call
+    // `FileSystemAccessHandleControlChild::ActorDestroy` which in turn calls
+    // `FileSystemSyncAccessHandle::ClearControlActor`, so `mControlActor`
+    // should be be null at this point.
+    MOZ_ASSERT(!mControlActor);
+  }
 }
 
 void FileSystemSyncAccessHandle::ClearActor() {
@@ -220,6 +249,14 @@
   mActor = nullptr;
 }
 
+void FileSystemSyncAccessHandle::ClearControlActor() {
+  // `mControlActor` is initialized in the constructor and this method is
+  // supposed to be called only once.
+  MOZ_ASSERT(mControlActor);
+
+  mControlActor = nullptr;
+}
+
 bool FileSystemSyncAccessHandle::IsOpen() const {
   MOZ_ASSERT(mState != State::Initial);
 